File: algo/quicksort.c
#include <stdio.h>

void exch(int* arr,int i,int j);

int partition(int* arr,int lo,int hi)
{
	int v=arr[lo];
	int i=lo;
	int j=hi+1;
	for(;;)
	{
		while(arr[++i]<v)
		{
			if(i>=hi)
				break;
		}
		while(arr[--j]>v)
		{
			if(j<=lo)
				break;
		}
		if(i>=j)
			break;
		exch(arr,i,j);
	}
	exch(arr,lo,j);
	return j;
}

void exch(int* arr,int i,int j)
{
	int t=arr[i];
	arr[i]=arr[j];
	arr[j]=t;
}

void qsort(int* arr,int lo,int hi)
{
	if(lo>=hi)
		return;
	int m=partition(arr,lo,hi);
	qsort(arr,lo,m-1);
	qsort(arr,m+1,hi);
}

int main()
{
#define LEN 100
	int arr[LEN];
	int i;
	for(i=0;i<LEN;++i)
	{
		arr[i]=LEN-i;
	}
	qsort(arr,0,LEN-1);
	for(i=0;i<LEN;++i)
		printf("%d,",arr[i]);

	printf("\n");
	return 0;
}

File: ruby/quicksort.rb
#! /usr/bin/ruby

def partition(arr,lo,hi)
	pivot=lo

	i=lo+1
	j=hi

	while true
		while true
			if i>=hi
				break
			end
			if arr[i]>=arr[pivot]
				break
			end
			i=i+1
		end

		while true
			if j<=lo 
				break
			end
			if arr[j]<=arr[pivot]
				break
			end
			j=j-1
		end

		if i>=j
			break
		end
		exch(arr,i,j)
	end

	exch(arr,j,pivot)
	return j
end

def exch(arr,i,j)
	tmp=arr[i]
	arr[i]=arr[j]
	arr[j]=tmp
end

def sort(arr,lo,hi)
	if
		lo>=hi
		return 
	end
	index=partition(arr,lo,hi)
	sort(arr,lo,index)
	sort(arr,index+1,hi)
end

def test
	#arr=10.times.map{Random.rand(1000)}
	arr=[2,1]
	puts "Before sort:"
	puts arr.to_s
	sort(arr,0,arr.size-1)
	puts "After sort:"
	puts arr.to_s
end

test
